Implement extended device filtering

Abyss777 opened the pull request:
There are three key changes:

1. Improve devices panel to use grid filters
- Removed bottom bar
- Added Status column
- Added filters for three columns: Name, Group, Status
- Increased panel width to not cut LastUpdate column by default
- Used `VisibleDevices` as chained store
2. Reflect device filtering on the map
- Set Id to all device related features to easily check feature showed or not
- Move taking decision of show or hide feature to `VisibleDevices` event handlers because only there we can make it
3. Save device panel state in cookies 
- There are a lot of possible convenient configurations of columns, I decided to store grid state to cookies, especially because it is really easy (just 5 additional lines)

Here is some screenshots
Default view
![image](https://cloud.githubusercontent.com/assets/5688080/24195913/6bddc446-0f1d-11e7-9a29-406065abdb67.png)

Name filter
![image](https://cloud.githubusercontent.com/assets/5688080/24195929/7f86d88e-0f1d-11e7-979c-85264f36aebd.png)

Group filter
![image](https://cloud.githubusercontent.com/assets/5688080/24195945/8ccb35f8-0f1d-11e7-83ab-2fb4675ab1fc.png)

Status filter
![image](https://cloud.githubusercontent.com/assets/5688080/24195966/9e74a06e-0f1d-11e7-992b-b0412cd51aca.png)

Cookies in store
![image](https://cloud.githubusercontent.com/assets/5688080/24195994/bee99804-0f1d-11e7-92fa-96e19c91d608.png)

More comments in code

Abyss777 commented on web/app/view/Devices.js:
```diff
@@ -20,20 +20,19 @@ Ext.define('Traccar.view.Devices', {
+    store: 'VisibleDevices',
+
+    stateful: true,
+    stateId: 'devices-grid',
```
Enable state autosave and set id

Abyss777 commented on web/app/view/Devices.js:
```diff
@@ -85,79 +84,30 @@ Ext.define('Traccar.view.Devices', {
     columns: {
         defaults: {
-            flex: 1,
+            flex: 2,
```
All columns are likely more wide than Status

Abyss777 commented on web/app/view/Devices.js:
```diff
@@ -169,26 +119,46 @@ Ext.define('Traccar.view.Devices', {
+                labelField: 'name',
+                store: 'DeviceStatuses'
+            },
+            renderer: function (value, metaData) {
```
This renderer and filter relies on https://github.com/tananaev/traccar/pull/3034

Abyss777 commented on web/app/view/MapMarkerController.js:
```diff
@@ -198,11 +204,11 @@ Ext.define('Traccar.view.MapMarkerController', {
                 feature = new ol.Feature(new ol.geom.Circle(center, radius));
                 feature.setStyle(this.getAreaStyle(null, Traccar.Style.mapAccuracyColor));
-                this.getView().getAccuracySource().addFeature(feature);
+                feature.setId(position.get('deviceId'));
```
Set feature id as device id (and for other device related features)

Abyss777 commented on web/app/view/MapMarkerController.js:
```diff
@@ -488,5 +494,56 @@ Ext.define('Traccar.view.MapMarkerController', {
+        accuracy = this.accuracyCircles[deviceId];
+        liveLine = this.liveRoutes[deviceId];
+        if (Ext.getStore('VisibleDevices').contains(device)) {
+            if (marker && !this.getView().getMarkersSource().getFeatureById(marker.getId())) {
```
Add feature only if it is not already there

tananaev replied:
What would happen if we call add if it's already there? Would it crash or OpenLayers would check it internally? If latter, it would be more readable to avoid this check.

Abyss777 replied:
It throws this exception https://openlayers.org/en/v4.0.0/doc/errors/#30 if feature already there
and this then try to remove not existed feature
![image](https://cloud.githubusercontent.com/assets/5688080/24232304/d8b18aa6-0fab-11e7-9cbd-f40ea0a51513.png)

Abyss777 commented on web/app/view/MapMarkerController.js:
```diff
@@ -488,5 +494,56 @@ Ext.define('Traccar.view.MapMarkerController', {
+                this.getView().getLiveRouteSource().addFeature(liveLine);
+            }
+        } else {
+            if (marker && this.getView().getMarkersSource().getFeatureById(marker.getId())) {
```
Remove feature only if it is there

tananaev replied:
Same question... can we remove feature if it's not there? Would it crash?

Abyss777 commented on web/app/view/MapMarkerController.js:
```diff
@@ -488,5 +494,56 @@ Ext.define('Traccar.view.MapMarkerController', {
+                this.getView().getMarkersSource().removeFeature(marker);
+            }
+            if (this.selectedMarker && marker && marker.getId() === this.selectedMarker.getId()) {
+                this.deselectFeature();
```
Deselect feature if need to hide it

tananaev replied:
Wouldn't device grid do it automatically?

Abyss777 commented on web/app/view/MapMarkerController.js:
```diff
@@ -488,5 +494,56 @@ Ext.define('Traccar.view.MapMarkerController', {
+        }
+    },
+
+    filterDevices: function (store) {
```
Handle filter changing here

Abyss777 commented on web/app/view/Devices.js:
```diff
@@ -169,24 +119,46 @@ Ext.define('Traccar.view.Devices', {
+                store: 'DeviceStatuses'
+            },
+            renderer: function (value, metaData) {
+                if (value) {
```
It's turned out that rendering happens twice after device was added. First time with empty string, second time with value from server response (offline).

Abyss777 commented:
How it looks on mobile
![image](https://cloud.githubusercontent.com/assets/5688080/24198019/e01bc7b0-0f25-11e7-94c3-cb13c30e47e0.png)

tananaev commented on web/app/controller/Root.js:
```diff
@@ -25,6 +25,10 @@ Ext.define('Traccar.controller.Root', {
     ],
 
+    init: function () {
+        Ext.state.Manager.setProvider(new Ext.state.CookieProvider());
```
This is amazing. Can it save status of everything else as well (e.g. buttons on the map)? This way we won't need to save them in the user attributes.

Abyss777 replied:
Looks like we can http://docs.sencha.com/extjs/6.2.0/classic/Ext.button.Button.html#cfg-stateful

tananaev commented on web/app/store/DeviceStatuses.js:
```diff
@@ -0,0 +1,35 @@
+    data: [{
+        id: 'online',
+        name: Strings.deviceStatusOnline,
+        tdCls: 'view-color-green'
```
I think we can use a better name instead of `tdCls`. Maybe just `style` or something.

tananaev commented on web/app/view/Devices.js:
```diff
@@ -169,24 +119,46 @@ Ext.define('Traccar.view.Devices', {
+            },
+            renderer: function (value, metaData) {
+                if (value) {
+                    metaData.tdCls = Ext.getStore('DeviceStatuses').getById(value).get('tdCls');
```
I think it would be better to call `getById` just once.

tananaev commented on web/app/view/MapMarkerController.js:
```diff
@@ -488,5 +494,56 @@ Ext.define('Traccar.view.MapMarkerController', {
         }
+    },
+
+    updateVisibleDevice: function (store, data) {
```
Does this method ever gets array? I guess then it would make sense to rename it to plural "devices".

tananaev commented on web/l10n/en.json:
```diff
@@ -84,6 +84,11 @@
     "deviceFollow": "Follow",
     "deviceTotalDistance": "Total Distance",
+    "deviceStatus": "Status",
+    "deviceSearchFor": "Search for...",
```
Do we really need this message? I think it should be pretty obvious even if the field is empty.

Abyss777 replied:
Unfortunately it is not empty
![image](https://cloud.githubusercontent.com/assets/5688080/24233579/fdc4739a-0fb4-11e7-9d5f-2dd91bf338fd.png)

Abyss777 commented:
Fixed all your comments, except `get*Source().getFeatureById()` check, unfortunately we can't avoid it, but I hope "byId" does not cost a lot.

Also fixed one missed issue, show features on `updateLatest` as soon as possible and in more correct way, not waiting next `updateDevice` event.

Also added `allowNull: true` to string fields of Device model, If need I will add it to all other models in other pull request.

Abyss777 commented:
Tomorrow is a new day.
I was thinking... We can implement some flag inside features itself to know if it is visible or not instead of checking `get*Source().getFeatureById()`. It might be faster. What do you think?

tananaev commented:
Merged the code, thanks.

I think we can leave the check there for now. If we find any performance issues, we can always implement optimization later.
